paz/models/detection/frustum_pointnet.py

```diff
@@ -0,0 +1,177 @@
+
+    net = Conv2D(128, kernel_size=[1, 1], padding='VALID', strides=[1, 1], activation='relu', name='conv2_2')(net)
+
+    net = Conv2D(256, kernel_size=[1, 1], padding='VALID', strides=[1, 1], activation='relu', name='conv2_3')(net)
+
+    net = Conv2D(512, kernel_size=[1, 1], padding='VALID', strides=[1, 1], activation='relu', name='conv2_4')(net)
+
+    net = MaxPooling2D(pool_size=[num_point, 1], padding='VALID')(net)
+
+    net = tf.squeeze(net, axis=[1, 2])
+    net = tf.concat([net, one_hot_vec], axis=1)
+
+    net = Dense(units=512, activation='relu')(net)
+    net = Dense(units=256, activation='relu')(net)
+
+    output = Dense(units=3 + NUM_HEADING_BIN * 2 + NUM_SIZE_CLUSTER * 4, activation=None)(net)
+
+    return output
+
+
+def TNet(object_point_cloud, one_hot_vec):
+    num_point = object_point_cloud.get_shape()[1]
+    net = tf.expand_dims(object_point_cloud, 2)
+
+    net = Conv2D(128, kernel_size=[1, 1], padding='VALID', strides=[1, 1], activation='relu', name='conv3_1')(net)
+
+    net = Conv2D(128, kernel_size=[1, 1], padding='VALID', strides=[1, 1], activation='relu', name='conv3_2')(net)
+
+    net = Conv2D(256, kernel_size=[1, 1], padding='VALID', strides=[1, 1], activation='relu', name='conv3_3')(net)
+
+    net = MaxPooling2D(pool_size=[num_point, 1], padding='VALID')(net)
+
+    net = tf.squeeze(net, axis=[1, 2])
+    net = tf.concat([net, one_hot_vec], axis=1)
+
+    net = Dense(units=256, activation='relu')(net)
+    net = Dense(units=128, activation='relu')(net)
+    predicted_center = Dense(units=3, activation=None)(net)
+
+    return predicted_center
+
+
+def FrustumPointNetModel(point_cloud_shape=(1024, 3), one_hot_vec_shape=(3,), mask_label_shape=(1024,),
+                         center_label_shape=(3,), heading_class_label_shape=(), heading_residual_label_shape=(),
+                         size_class_label_shape=(), size_residual_label_shape=(3,), batch_size=32):
+    end_points = {}
+
+    point_cloud = Input(point_cloud_shape, name="frustum_point_cloud", batch_size=batch_size)
+    one_hot_vec = Input(one_hot_vec_shape, name="one_hot_vec", batch_size=batch_size)
+    mask_label = Input(mask_label_shape, name="seg_label", batch_size=batch_size)
+    center_label = Input(center_label_shape, name="box3d_center", batch_size=batch_size)
+    heading_class_label = Input(heading_class_label_shape, name="angle_class", batch_size=batch_size)
+    heading_residual_label = Input(heading_residual_label_shape, name="angle_residual", batch_size=batch_size)
+    size_class_label = Input(size_class_label_shape, name="size_class", batch_size=batch_size)
+    size_residual_label = Input(size_residual_label_shape, name="size_residual", batch_size=batch_size)
+
+    logits = Instance_Seg_Net(point_cloud, one_hot_vec)  # bs,n,2
+    end_points['mask_logits'] = logits
+
+    # Mask Point Centroid
+    object_point_cloud_xyz, mask_xyz_mean, end_points = point_cloud_masking(point_cloud, logits, end_points)
```

Comment:
Maybe it would be cleaner if you wrap "point_cloud_masking" as a layer rather than a function.